### ISSUE TITLE
Add a `perform` method

### ISSUE THREAD

**Issue by robb** [robb/Underscore.m#3]
Add a `perform` method similar to [underscore.js' invoke method](http://documentcloud.github.com/underscore/#invoke) that performs a selector on every object in the array/dictionary.

**Comment by robb** [comment]
Here is what I came up with:

```
- (USArrayWrapper *(^)(SEL))perform;
{
    return ^USArrayWrapper *(SEL selector) {
        return self.each(^(id obj) {
            if ([obj respondsToSelector:selector]) {
                [obj performSelector:selector];
            }
        });
    };
}
```

This, however, yields a warning `performSelector may cause a leak because its selector is unknown` because the compiler does not know how to handle the return value of `performSelector:` with respect to ARC. (I.e. should it be released or not?)

I could suppress the warning by using

```
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
...
#pragma clang diagnostic pop
```

but this will open up the opportunity to leak return values.
Obvious example would be `_array(stuff).perform(@selector(copy)).unwrap`

Any opinions on this? Does the end justify the means?

**Comment by robb** [comment]
The more I think about this, the more I lean towards doing it, with a warning in the comments.

**Comment by m0wfo** [comment]
I'm not at my Mac right now, but would creating an NSInvocation be a workaround? What if the user wants to supply a list of args as well as a selector?

**Comment by robb** [comment]
You're right, using an NSInvocation yields no such warning.

I don't think something like `performWith(sel, arg1, arg2, nil)` is really needed and would suggest using `each` instead, though I am open to reconsider that in the future if it becomes an apparent pattern :-)

**Comment by robb** [comment]
_Scratch that,_ I did some research and grepped the iOS 5.1 frameworks.
There seems to be only a small number of `void` methods that take no parameters. A perform without arguments makes probably not much sense then.

Maybe something like `invoke(NSInvocation *invocation)` instead?

/cc @stigi @toto

**Comment by Gernot** [comment]
You don't want the developer to create an NSInvocation object himself and supply it, because he'd have to hassle with all the MethodSignature stuff, etc. What you might do however is  taking the selector and creating an NSInvocation with the correct MethodSignature yourself. You have to assume that you only use objects as parameters, but -performSelector is the same in that regard.

**Comment by toto** [comment]
I would suggest keeping this similar to `-[NSArray makeObjectsPerformSelector]` and add a second method similar to `-[NSArray makeObjectsPerformSelectorWithObject:]` which only takes object arguments just like `performSelector:withObject:`.

**Comment by robb** [comment]
So it boils down to:

```
 perform(SEL selector, id arg, ...)
```

or

```
perform(SEL selector, NSArray *args)
```

or

```
perform(SEL selector)
perform1(SEL selector, id arg1)
perform2(SEL selector, id arg1, id arg2)
```

After some discussion with @toto and @Gernot, I'm leaning towards the 2nd option.
It may seem like a lot of typing now, but

```
perform(@selector(none), nil)
perform(@selector(foo:), [NSArray arrayWithObject:foo])
perform(@selector(foo:bar:), [NSArray arrayWithObjects:foo, bar, nil])
```

will, with Xcode 4.4, turn into

```
perform(@selector(none), nil)
perform(@selector(foo:), @[foo])
perform(@selector(foo:bar:), @[foo, bar])
```

Literals FTW!

**Comment by stigi** [comment]
What about using va_list for the arguments?

**Comment by toto** [comment]
Been there. Arrays should be easier with literally coming otherwise va+blocks seemed a bit to much.

**Comment by robb** [comment]
Just occurred to me that the selector exposes the number of arguments, so no nil termination is necessary.
That makes the varargs solution a lot more attractive.

**Comment by robb** [comment]
(Sorry, Github issues app had issues)

Anyways, the interface can totally look like

```
perform(@selector(none))
perform(@selector(foo:), foo)
perform(@selector(foo:bar:), foo, bar)
```

I do like having the ability to apply an Array of objects instead of a list of arguments when performing a selector, but that probably belongs to `USObjectWrapper` ;-)

This is also the only approach that lets you perform the selector with nil arguments, as could not put them into an argument array or use them with nil-termination.

**Comment by robb** [comment]
This is what I came up with:

```
- (USArrayWrapper *(^)(SEL, ...))perform
{
    return ^USArrayWrapper *(SEL selector, ...) {
        if (self.array.count == 0) {
            return self;
        }

        NSMethodSignature *methodSignature = [self.first methodSignatureForSelector:selector];
        NSInvocation      *invocation      = [NSInvocation invocationWithMethodSignature:methodSignature];
        invocation.selector = selector;

        va_list args;
        va_start(args, selector);
        for (NSUInteger i = 2; i < methodSignature.numberOfArguments; i++) {
            id arg = va_arg(args, id);
            [invocation setArgument:&arg atIndex:i];
        }
        va_end(args);

        return self.each(^(id obj) {
            [invocation setTarget:obj];
            [invocation invoke];
        });
    };
}
```

It makes the assumption that `methodSignatureForSelector:` will return a method with the same number of arguments for every object in the array, which is not necessarily true. Alternatively, one could create a new invocation for every object, but that doesn't sound too promising either.

Maybe not worth the effort after all? It's not like

```
_array(foo)
    .each(^(id obj) {
        [obj performSelector:foo];
    });
```

is super noisy.

**Comment by robb** [comment]
Closing this as a wontfix for now, feel free to reopen this if you come up with a solution we haven't considered.
I just don't want to have 0.1.0 blocked by it.
